File: src/main/java/jp/kusumotolab/kgenprog/CUILauncher.java
package jp.kusumotolab.kgenprog;

import java.util.Random;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ch.qos.logback.classic.Level;
import jp.kusumotolab.kgenprog.fl.FaultLocalization;
import jp.kusumotolab.kgenprog.ga.codegeneration.DefaultSourceCodeGeneration;
import jp.kusumotolab.kgenprog.ga.codegeneration.SourceCodeGeneration;
import jp.kusumotolab.kgenprog.ga.crossover.Crossover;
import jp.kusumotolab.kgenprog.ga.crossover.FirstVariantSelectionStrategy;
import jp.kusumotolab.kgenprog.ga.crossover.SecondVariantSelectionStrategy;
import jp.kusumotolab.kgenprog.ga.mutation.Mutation;
import jp.kusumotolab.kgenprog.ga.mutation.SimpleMutation;
import jp.kusumotolab.kgenprog.ga.mutation.selection.CandidateSelection;
import jp.kusumotolab.kgenprog.ga.mutation.selection.RouletteStatementAndConditionSelection;
import jp.kusumotolab.kgenprog.ga.selection.DefaultVariantSelection;
import jp.kusumotolab.kgenprog.ga.selection.VariantSelection;
import jp.kusumotolab.kgenprog.ga.validation.DefaultCodeValidation;
import jp.kusumotolab.kgenprog.ga.validation.SourceCodeValidation;
import jp.kusumotolab.kgenprog.output.Exporter;
import jp.kusumotolab.kgenprog.output.ExporterFactory;
import jp.kusumotolab.kgenprog.project.test.LocalTestExecutor;
import jp.kusumotolab.kgenprog.project.test.TestExecutor;

public class CUILauncher {

  public static void main(final String[] args) {
    try {
      final Configuration config = Configuration.Builder.buildFromCmdLineArgs(args);
      final CUILauncher launcher = new CUILauncher();
      launcher.launch(config);
    } catch (final RuntimeException e) {
      System.exit(1);
    }
  }

  public void launch(final Configuration config) {
    setLogLevel(config.getLogLevel());

    final FaultLocalization faultLocalization = config.getFaultLocalization()
        .initialize();
    final Random random = new Random(config.getRandomSeed());
    final CandidateSelection candidateSelection =
        new RouletteStatementAndConditionSelection(random);
    final Mutation mutation = new SimpleMutation(config.getMutationGeneratingCount(), random,
        candidateSelection, config.getScope());
    final FirstVariantSelectionStrategy firstVariantSelectionStrategy =
        config.getFirstVariantSelectionStrategy()
            .initialize(random);
    final SecondVariantSelectionStrategy secondVariantSelectionStrategy =
        config.getSecondVariantSelectionStrategy()
            .initialize(random);
    final Crossover crossover = config.getCrossoverType()
        .initialize(random, firstVariantSelectionStrategy,
            secondVariantSelectionStrategy, config.getCrossoverGeneratingCount());
    final SourceCodeGeneration sourceCodeGeneration = new DefaultSourceCodeGeneration();
    final SourceCodeValidation sourceCodeValidation = new DefaultCodeValidation();
    final VariantSelection variantSelection = new DefaultVariantSelection(config.getHeadcount(),
        random);
    final TestExecutor testExecutor = new LocalTestExecutor(config);
    final Exporter exporter = ExporterFactory.create(config);
    final KGenProgMain kGenProgMain =
        new KGenProgMain(config, faultLocalization, mutation, crossover, sourceCodeGeneration,
            sourceCodeValidation, variantSelection, testExecutor, exporter);

    kGenProgMain.run();
  }

  private void setLogLevel(final Level logLevel) {
    final ch.qos.logback.classic.Logger rootLogger =
        (ch.qos.logback.classic.Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);
    rootLogger.setLevel(logLevel);
  }
}

File: src/main/kotlin/jp/kusumotolab/kgenprog/CUILauncher.kt
package jp.kusumotolab.kgenprog

import jp.kusumotolab.kgenprog.fl.FaultLocalization
import jp.kusumotolab.kgenprog.fl.Ochiai
import jp.kusumotolab.kgenprog.ga.Crossover
import jp.kusumotolab.kgenprog.ga.DefaultCodeValidation
import jp.kusumotolab.kgenprog.ga.DefaultSourceCodeGeneration
import jp.kusumotolab.kgenprog.ga.DefaultVariantSelection
import jp.kusumotolab.kgenprog.ga.Mutation
import jp.kusumotolab.kgenprog.ga.RandomMutation
import jp.kusumotolab.kgenprog.ga.SiglePointCrossover
import jp.kusumotolab.kgenprog.ga.SourceCodeGeneration
import jp.kusumotolab.kgenprog.ga.SourceCodeValidation
import jp.kusumotolab.kgenprog.ga.VariantSelection
import jp.kusumotolab.kgenprog.project.ClassPath
import jp.kusumotolab.kgenprog.project.SourceFile
import jp.kusumotolab.kgenprog.project.TargetProject
import org.kohsuke.args4j.Argument
import org.kohsuke.args4j.CmdLineException
import org.kohsuke.args4j.CmdLineParser
import org.kohsuke.args4j.Option
import java.util.*

class CUILauncher {

    @Option(name = "-s", aliases = ["--src"], required = true, usage = "Paths of the root directories holding source codes")
    var sourceFiles = mutableListOf("src/main/java")
    @Option(name = "-t", aliases = ["--test"], required = true, usage = "Paths of the root directories holding test codes")
    var testFiles = mutableListOf<String>()
    @Option(name = "-c", aliases = ["--cp"], required = true, usage = "Class paths required to build the target project")
    var classPaths = mutableListOf<String>()

    fun launch() {
        val sourceFiles = this.sourceFiles
                .map { SourceFile(it) }
                .toList()

        val testFiles = this.testFiles
                .map { SourceFile(it) }
                .toList()

        val classPaths = this.classPaths
                .map { ClassPath(it) }
                .toList()

        val targetProject = TargetProject(sourceFiles, testFiles, classPaths)
        val faultLocalization: FaultLocalization = Ochiai()
        val mutation: Mutation = RandomMutation()
        val crossover: Crossover = SiglePointCrossover()
        val sourceCodeGeneration: SourceCodeGeneration = DefaultSourceCodeGeneration()
        val sourceCodeValidation: SourceCodeValidation = DefaultCodeValidation()
        val variantSelection: VariantSelection = DefaultVariantSelection()

        val kGenProgMain = KGenProgMain(targetProject, faultLocalization, mutation, crossover, sourceCodeGeneration, sourceCodeValidation, variantSelection)
        kGenProgMain.run()
    }

    companion object {


        @JvmStatic
        fun main(args: Array<String>) {
            val launcher = CUILauncher()
            val parser = CmdLineParser(launcher)

            try {
                parser.parseArgument(*args)
            } catch (e: CmdLineException) {
                System.err.println(e.message)
                parser.printUsage(System.err)
                System.exit(1)
            }
            launcher.launch()
        }
    }
}
